src/Builder/Encoder/AbstractExpressionEncoder.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MongoDB\Builder\Encoder;
+
+use MongoDB\Builder\BuilderEncoder;
+use stdClass;
+
+use function get_object_vars;
+use function is_array;
+
+/**
+ * @template BSONType of stdClass|array|string
+ * @template NativeType
+ * @template-implements ExpressionEncoder<BSONType, NativeType>
+ */
+abstract class AbstractExpressionEncoder implements ExpressionEncoder
+{
+    final public function __construct(protected readonly BuilderEncoder $encoder)
+    {
+    }
+
+    /**
+     * Nested arrays and objects must be encoded recursively.
+     *
+     * @psalm-param T $value
+     *
+     * @psalm-return (T is object ? object : (T is array ? array : mixed))
```

Comment:
Should this use `stdClass` instead of `object`? [Object Types](https://psalm.dev/docs/annotating_code/type_syntax/object_types/) suggests that those are distinct.

That said, the code will still end up returning a non-stdClass object as-is, which won't invalidate the this annotation.

Separate question: why return `mixed` instead of `T` for the non-object, non-array case? [`@template`, `@psalm-template`](https://psalm.dev/docs/annotating_code/templated_annotations/#template-psalm-template) has an example of doing so, and I assume all extending classes will substitute `T` (correct me if I'm mistaken).